File: koduck-frontend/src/api/portfolio.ts
import { apiClient } from './client';

// ============================================================================
// Legacy Types (for backward compatibility with existing Portfolio page)
// ============================================================================

export interface PortfolioItem {
  id: string;
  symbol: string;
  name: string;
  sector: string;
  quantity: number;
  avgCost: number;
  currentPrice: number;
  marketValue: number;
  pnl: number;
  pnlPercent: number;
  weight: number;
}

export interface PortfolioSummary {
  totalValue: number;
  totalCost: number;
  totalPnl: number;
  totalPnlPercent: number;
  dailyPnl: number;
  dailyPnlPercent: number;
  availableCash: number;
  totalEquity: number;
}

export interface TradeRecord {
  id: string;
  symbol: string;
  name: string;
  type: 'buy' | 'sell';
  quantity: number;
  price: number;
  amount: number;
  time: string;
  date: string;
}

export interface SectorDistribution {
  name: string;
  value: number;
  percent: number;
  color?: string;
}

// Legacy API object for backward compatibility
export const portfolioApi = {
  async getPortfolio(): Promise<PortfolioItem[]> {
    const response = await apiClient.get('/portfolio/positions');
    return response || [];
  },

  async getPortfolioSummary(): Promise<PortfolioSummary> {
    const response = await apiClient.get('/portfolio/summary');
    return response || {
      totalValue: 0,
      totalCost: 0,
      totalPnl: 0,
      totalPnlPercent: 0,
      dailyPnl: 0,
      dailyPnlPercent: 0,
      availableCash: 0,
      totalEquity: 0,
    };
  },

  async getTradeRecords(): Promise<TradeRecord[]> {
    const response = await apiClient.get('/portfolio/trades');
    return response || [];
  },

  async getSectorDistribution(positions?: PortfolioItem[]): Promise<SectorDistribution[]> {
    // If positions provided, calculate distribution client-side
    if (positions && positions.length > 0) {
      const sectorMap = new Map<string, number>();
      const totalValue = positions.reduce((sum, p) => sum + p.marketValue, 0);

      positions.forEach((position) => {
        const current = sectorMap.get(position.sector) || 0;
        sectorMap.set(position.sector, current + position.marketValue);
      });

      const colors = ['#00F2FF', '#DE0541', '#FFB3B5', '#FFD81D', '#00DBE7', '#7D7D7D'];

      return Array.from(sectorMap.entries()).map(([name, value], index) => ({
        name,
        value,
        percent: totalValue > 0 ? (value / totalValue) * 100 : 0,
        color: colors[index % colors.length],
      }));
    }

    // Otherwise fetch from API
    const response = await apiClient.get('/portfolio/sector-distribution');
    return response || [];
  },
};

// ============================================================================
// Types - Issue #203: Northbound Flow
// ============================================================================

export interface PeriodFlow {
  name: string;
  inflow: number;
  inflow_formatted: string;
  start_time: string;
  end_time: string;
}

export interface StockFlow {
  symbol: string;
  name: string;
  net_flow: number;
  buy_amount: number;
  sell_amount: number;
  holding_change: number;
}

export interface NorthboundFlowResponse {
  date: string;
  total_inflow: number;
  total_inflow_formatted: string;
  shanghai_inflow: number;
  shenzhen_inflow: number;
  periods: PeriodFlow[];
  top_buys: StockFlow[];
  top_sells: StockFlow[];
  cumulative_inflow_5d: number;
  cumulative_inflow_20d: number;
  timestamp: string;
}

export interface NorthboundHistoryItem {
  date: string;
  inflow: number;
  inflow_formatted: string;
  shanghai_inflow: number;
  shenzhen_inflow: number;
}

export interface NorthboundHistoryResponse {
  data: NorthboundHistoryItem[];
  total_days: number;
}

// ============================================================================
// Types - Issue #208: PnL History
// ============================================================================

export interface PnLDataPoint {
  date: string;
  timestamp: number;
  pnl: number;
  pnl_formatted: string;
  pnl_percent: number;
  total_cost: number;
  total_market_value: number;
  cumulative_pnl: number;
}

export interface PnLSummary {
  total_pnl: number;
  total_pnl_formatted: string;
  total_pnl_percent: number;
  best_day?: PnLDataPoint;
  worst_day?: PnLDataPoint;
  volatility: number;
  sharpe_ratio: number;
}

export interface PnLHistoryResponse {
  period: string;
  start_date: string;
  end_date: string;
  data: PnLDataPoint[];
  summary: PnLSummary;
  benchmark_comparison: {
    portfolio_return: number;
    benchmark_return: number;
    alpha: number;
    beta: number;
  };
}

export type PeriodType = '1d' | '1w' | '1m' | '3m' | '1y' | 'ytd';

// ============================================================================
// Types - Issue #209: Sector Allocation
// ============================================================================

export interface SectorAllocationItem {
  name: string;
  code: string;
  value: number;
  value_formatted: string;
  percent: number;
  color: string;
  stock_count: number;
  stocks: Array<{
    symbol: string;
    name: string;
    value: number;
  }>;
}

export interface SectorAllocationResponse {
  total_value: number;
  total_value_formatted: string;
  cash_value: number;
  cash_percent: number;
  sectors: SectorAllocationItem[];
  diversification_score: number;
  top_heavy_risk: 'low' | 'medium' | 'high';
}

// ============================================================================
// API Functions - Issue #203: Northbound Flow
// ============================================================================

export async function getNorthboundFlow(tradeDate?: string): Promise<NorthboundFlowResponse> {
  const response = await apiClient.get('/market/northbound-flow', {
    params: { trade_date: tradeDate }
  });
  return response;
}

export async function getNorthboundHistory(days: number = 30): Promise<NorthboundHistoryResponse> {
  const response = await apiClient.get('/market/northbound-flow/history', {
    params: { days }
  });
  return response;
}

export async function getNorthboundStats(): Promise<any> {
  const response = await apiClient.get('/market/northbound-flow/stats');
  return response;
}

// ============================================================================
// API Functions - Issue #208: PnL History
// ============================================================================

export async function getPnLHistory(period: PeriodType = '1w'): Promise<PnLHistoryResponse> {
  const response = await apiClient.get('/portfolio/pnl-history', {
    params: { period }
  });
  return response;
}

export async function getDailyPnL(startDate?: string, endDate?: string): Promise<PnLDataPoint[]> {
  const response = await apiClient.get('/portfolio/pnl-history/daily', {
    params: { start_date: startDate, end_date: endDate }
  });
  return response;
}

// ============================================================================
// API Functions - Issue #209: Sector Allocation
// ============================================================================

export async function getSectorAllocation(minPercent: number = 0.5): Promise<SectorAllocationResponse> {
  const response = await apiClient.get('/portfolio/sector-allocation', {
    params: { min_percent: minPercent }
  });
  return response;
}

export async function getSectorAllocationTrend(months: number = 6): Promise<any[]> {
  const response = await apiClient.get('/portfolio/sector-allocation/trend', {
    params: { months }
  });
  return response;
}

export async function getPortfolioHoldings(sortBy: string = 'value', limit: number = 50): Promise<any[]> {
  const response = await apiClient.get('/portfolio/holdings', {
    params: { sort_by: sortBy, limit }
  });
  return response;
}

// ============================================================================
// Mock Data
// ============================================================================

export const mockNorthboundFlow: NorthboundFlowResponse = {
  date: '2024-01-15',
  total_inflow: 2400000000,
  total_inflow_formatted: '+¥24.0亿',
  shanghai_inflow: 1440000000,
  shenzhen_inflow: 960000000,
  periods: [
    { name: 'OPEN', inflow: 800000000, inflow_formatted: '+¥8.0亿', start_time: '09:30', end_time: '10:30' },
    { name: 'MID-DAY', inflow: 1200000000, inflow_formatted: '+¥12.0亿', start_time: '10:30', end_time: '14:00' },
    { name: 'CLOSE', inflow: 400000000, inflow_formatted: '+¥4.0亿', start_time: '14:00', end_time: '15:00' },
  ],
  top_buys: [
    { symbol: '600519', name: '贵州茅台', net_flow: 523000000, buy_amount: 892000000, sell_amount: 369000000, holding_change: 0.12 },
    { symbol: '000858', name: '五粮液', net_flow: 312000000, buy_amount: 456000000, sell_amount: 144000000, holding_change: 0.18 },
  ],
  top_sells: [
    { symbol: '002594', name: '比亚迪', net_flow: -234000000, buy_amount: 123000000, sell_amount: 357000000, holding_change: -0.14 },
  ],
  cumulative_inflow_5d: 8910000000,
  cumulative_inflow_20d: 34500000000,
  timestamp: new Date().toISOString()
};

export const mockPnLHistory: PnLHistoryResponse = {
  period: '1w',
  start_date: '2024-01-08',
  end_date: '2024-01-15',
  data: [
    { date: '2024-01-08', timestamp: 1704672000, pnl: 120000, pnl_formatted: '+12.0万', pnl_percent: 2.4, total_cost: 5000000, total_market_value: 5120000, cumulative_pnl: 120000 },
    { date: '2024-01-09', timestamp: 1704758400, pnl: 50000, pnl_formatted: '+5.0万', pnl_percent: 2.5, total_cost: 5000000, total_market_value: 5170000, cumulative_pnl: 170000 },
    { date: '2024-01-10', timestamp: 1704844800, pnl: -30000, pnl_formatted: '-3.0万', pnl_percent: 2.44, total_cost: 5000000, total_market_value: 5140000, cumulative_pnl: 140000 },
    { date: '2024-01-11', timestamp: 1704931200, pnl: 80000, pnl_formatted: '+8.0万', pnl_percent: 2.6, total_cost: 5000000, total_market_value: 5220000, cumulative_pnl: 220000 },
    { date: '2024-01-12', timestamp: 1705017600, pnl: 45000, pnl_formatted: '+4.5万', pnl_percent: 2.69, total_cost: 5000000, total_market_value: 5265000, cumulative_pnl: 265000 },
    { date: '2024-01-15', timestamp: 1705276800, pnl: 65000, pnl_formatted: '+6.5万', pnl_percent: 2.82, total_cost: 5000000, total_market_value: 5330000, cumulative_pnl: 330000 },
  ],
  summary: {
    total_pnl: 330000,
    total_pnl_formatted: '+33.0万',
    total_pnl_percent: 6.6,
    best_day: { date: '2024-01-08', timestamp: 1704672000, pnl: 120000, pnl_formatted: '+12.0万', pnl_percent: 2.4, total_cost: 5000000, total_market_value: 5120000, cumulative_pnl: 120000 },
    worst_day: { date: '2024-01-10', timestamp: 1704844800, pnl: -30000, pnl_formatted: '-3.0万', pnl_percent: 2.44, total_cost: 5000000, total_market_value: 5140000, cumulative_pnl: 140000 },
    volatility: 5.2,
    sharpe_ratio: 1.27
  },
  benchmark_comparison: {
    portfolio_return: 6.6,
    benchmark_return: 5.1,
    alpha: 1.5,
    beta: 0.95
  }
};

export const mockSectorAllocation: SectorAllocationResponse = {
  total_value: 2841902,
  total_value_formatted: '284.2万',
  cash_value: 500000,
  cash_percent: 17.6,
  sectors: [
    { name: '食品饮料', code: '食品饮料', value: 850000, value_formatted: '85.0万', percent: 29.9, color: '#00F2FF', stock_count: 2, stocks: [{ symbol: '600519', name: '贵州茅台', value: 500000 }, { symbol: '000858', name: '五粮液', value: 350000 }] },
    { name: '医药生物', code: '医药生物', value: 620000, value_formatted: '62.0万', percent: 21.8, color: '#DE0541', stock_count: 2, stocks: [{ symbol: '600276', name: '恒瑞医药', value: 320000 }, { symbol: '300760', name: '迈瑞医疗', value: 300000 }] },
    { name: '电子', code: '电子', value: 480000, value_formatted: '48.0万', percent: 16.9, color: '#7D7D7D', stock_count: 2, stocks: [{ symbol: '002594', name: '比亚迪', value: 280000 }, { symbol: '300750', name: '宁德时代', value: 200000 }] },
    { name: '银行', code: '银行', value: 320000, value_formatted: '32.0万', percent: 11.3, color: '#00DBE7', stock_count: 2, stocks: [{ symbol: '600036', name: '招商银行', value: 200000 }] },
    { name: '电力设备', code: '电力设备', value: 210000, value_formatted: '21.0万', percent: 7.4, color: '#FFD81D', stock_count: 2, stocks: [{ symbol: '601012', name: '隆基绿能', value: 150000 }] },
    { name: '现金', code: 'CASH', value: 500000, value_formatted: '50.0万', percent: 17.6, color: '#64748B', stock_count: 0, stocks: [] },
  ],
  diversification_score: 72.5,
  top_heavy_risk: 'medium'
};

// Legacy mock data for Portfolio API
export const mockPortfolioItems: PortfolioItem[] = [
  { id: '1', symbol: '600519', name: '贵州茅台', sector: '食品饮料', quantity: 100, avgCost: 1500, currentPrice: 1650, marketValue: 165000, pnl: 15000, pnlPercent: 10, weight: 25 },
  { id: '2', symbol: '000858', name: '五粮液', sector: '食品饮料', quantity: 200, avgCost: 140, currentPrice: 155, marketValue: 31000, pnl: 3000, pnlPercent: 10.7, weight: 15 },
];

export const mockPortfolioSummary: PortfolioSummary = {
  totalValue: 500000,
  totalCost: 450000,
  totalPnl: 50000,
  totalPnlPercent: 11.1,
  dailyPnl: 5000,
  dailyPnlPercent: 1.0,
  availableCash: 100000,
  totalEquity: 600000,
};

export const mockTradeRecords: TradeRecord[] = [
  { id: '1', symbol: '600519', name: '贵州茅台', type: 'buy', quantity: 100, price: 1500, amount: 150000, time: '09:30:00', date: '2024-01-15' },
  { id: '2', symbol: '000858', name: '五粮液', type: 'buy', quantity: 200, price: 140, amount: 28000, time: '10:15:00', date: '2024-01-15' },
];

File: koduck-data-service/app/routers/portfolio.py
"""Portfolio API routes.

Issue #208: Portfolio 历史盈亏趋势 API
Issue #209: Portfolio 行业配置分布 API
"""

import logging
from typing import List, Optional
from datetime import datetime, date, timedelta, timezone
from enum import Enum

from fastapi import APIRouter, Query, HTTPException, Depends
from pydantic import BaseModel, Field

from app.models.schemas import ApiResponse

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/portfolio", tags=["portfolio"])


# ============================================================================
# Issue #208: PnL History Models
# ============================================================================

class PeriodType(str, Enum):
    """Time period for PnL history."""
    ONE_DAY = "1d"
    ONE_WEEK = "1w"
    ONE_MONTH = "1m"
    THREE_MONTH = "3m"
    ONE_YEAR = "1y"
    YTD = "ytd"


class PnLDataPoint(BaseModel):
    """Single PnL data point."""
    date: str = Field(..., description="Date string")
    timestamp: int = Field(..., description="Unix timestamp")
    pnl: float = Field(..., description="Profit/Loss amount")
    pnl_formatted: str = Field(..., description="Formatted PnL")
    pnl_percent: float = Field(..., description="PnL percentage")
    total_cost: float = Field(..., description="Total cost basis")
    total_market_value: float = Field(..., description="Total market value")
    cumulative_pnl: float = Field(..., description="Cumulative PnL")


class PnLSummary(BaseModel):
    """PnL summary statistics."""
    total_pnl: float = Field(..., description="Total profit/loss")
    total_pnl_formatted: str = Field(..., description="Formatted total PnL")
    total_pnl_percent: float = Field(..., description="Total PnL percentage")
    best_day: Optional[PnLDataPoint] = Field(None, description="Best performing day")
    worst_day: Optional[PnLDataPoint] = Field(None, description="Worst performing day")
    volatility: float = Field(..., description="PnL volatility (std dev)")
    sharpe_ratio: float = Field(..., description="Risk-adjusted return")


class PnLHistoryResponse(BaseModel):
    """PnL history response."""
    period: str = Field(..., description="Query period")
    start_date: str = Field(..., description="Start date")
    end_date: str = Field(..., description="End date")
    data: List[PnLDataPoint] = Field(..., description="PnL data points")
    summary: PnLSummary = Field(..., description="Summary statistics")
    benchmark_comparison: dict = Field(default={}, description="Benchmark comparison")


# ============================================================================
# Issue #209: Sector Allocation Models
# ============================================================================

class SectorAllocationItem(BaseModel):
    """Individual sector allocation."""
    name: str = Field(..., description="Sector name")
    code: str = Field(..., description="Sector code")
    value: float = Field(..., description="Market value")
    value_formatted: str = Field(..., description="Formatted value")
    percent: float = Field(..., description="Allocation percentage")
    color: str = Field(..., description="Display color")
    stock_count: int = Field(..., description="Number of stocks in sector")
    stocks: List[dict] = Field(default=[], description="Top stocks in sector")


class SectorAllocationResponse(BaseModel):
    """Sector allocation response."""
    total_value: float = Field(..., description="Total portfolio value")
    total_value_formatted: str = Field(..., description="Formatted total value")
    cash_value: float = Field(..., description="Cash position")
    cash_percent: float = Field(..., description="Cash percentage")
    sectors: List[SectorAllocationItem] = Field(..., description="Sector allocations")
    diversification_score: float = Field(..., description="Diversification score (0-100)")
    top_heavy_risk: str = Field(..., description="Concentration risk level")


class IndustryMapping:
    """Industry name mapping and colors."""
    
    SECTOR_COLORS = {
        "食品饮料": "#00F2FF",
        "医药生物": "#DE0541",
        "电子": "#FFB3B5",
        "电力设备": "#FFD81D",
        "银行": "#00DBE7",
        "非银金融": "#7D7D7D",
        "计算机": "#9CA3AF",
        "汽车": "#F59E0B",
        "机械设备": "#10B981",
        "化工": "#8B5CF6",
        "家用电器": "#EC4899",
        "有色金属": "#F97316",
        "建筑材料": "#64748B",
        "房地产": "#94A3B8",
        "交通运输": "#06B6D4",
        "通信": "#3B82F6",
        "传媒": "#A855F7",
        "国防军工": "#DC2626",
        "公用事业": "#059669",
        "煤炭": "#374151",
        "石油石化": "#1F2937",
        "钢铁": "#4B5563",
        "农林牧渔": "#16A34A",
        "商贸零售": "#D946EF",
        "社会服务": "#8B5CF6",
        "美容护理": "#F43F5E",
        "轻工制造": "#84CC16",
        "纺织服饰": "#14B8A6",
        "环保": "#22C55E",
        "综合": "#6B7280",
    }
    
    DEFAULT_COLOR = "#1D2026"
    
    @classmethod
    def get_color(cls, sector_name: str) -> str:
        return cls.SECTOR_COLORS.get(sector_name, cls.DEFAULT_COLOR)


# ============================================================================
# Mock Data Generators
# ============================================================================

def generate_pnl_history(period: str, days: int) -> List[PnLDataPoint]:
    """Generate mock PnL history data."""
    data = []
    end_date = date.today()
    
    # Starting values
    base_cost = 5000000  # 500万成本
    current_value = base_cost
    cumulative_pnl = 0
    
    for i in range(days, -1, -1):
        current_date = end_date - timedelta(days=i)
        if current_date.weekday() >= 5:  # Skip weekends
            continue
            
        # Generate daily PnL with some randomness but trending
        daily_change = (hash(str(current_date)) % 1000000) - 400000  # -40万到+60万
        if i > days // 2:
            daily_change += 50000  # Slight upward trend in first half
        
        current_value += daily_change
        cumulative_pnl += daily_change
        
        pnl_percent = ((current_value - base_cost) / base_cost) * 100
        
        data.append(PnLDataPoint(
            date=current_date.isoformat(),
            timestamp=int(datetime.combine(current_date, datetime.min.time()).timestamp()),
            pnl=daily_change,
            pnl_formatted=f"+{daily_change/10000:.1f}万" if daily_change >= 0 else f"{daily_change/10000:.1f}万",
            pnl_percent=pnl_percent,
            total_cost=base_cost,
            total_market_value=current_value,
            cumulative_pnl=cumulative_pnl
        ))
    
    return data


# ============================================================================
# API Endpoints - Issue #208
# ============================================================================

@router.get("/pnl-history", response_model=ApiResponse[PnLHistoryResponse])
async def get_pnl_history(
    period: PeriodType = Query(PeriodType.ONE_WEEK, description="Time period: 1d, 1w, 1m, 3m, 1y, ytd"),
    user_id: Optional[int] = Query(None, description="User ID (for admin/multi-user)")
):
    """Get portfolio PnL (Profit/Loss) history.
    
    Issue #208: Returns historical profit/loss data for portfolio trend chart.
    Supports multiple time periods: 1D, 1W, 1M, 3M, 1Y, YTD.
    
    Args:
        period: Time period for history data
        user_id: Optional user ID for multi-user systems
        
    Returns:
        PnLHistoryResponse with data points and summary statistics.
    """
    try:
        # Map period to days
        period_days = {
            PeriodType.ONE_DAY: 1,
            PeriodType.ONE_WEEK: 7,
            PeriodType.ONE_MONTH: 30,
            PeriodType.THREE_MONTH: 90,
            PeriodType.ONE_YEAR: 252,  # Trading days
            PeriodType.YTD: (date.today() - date(date.today().year, 1, 1)).days
        }
        
        days = period_days.get(period, 7)
        
        # Generate mock data
        data = generate_pnl_history(period.value, days)
        
        if not data:
            raise HTTPException(status_code=404, detail="No data available for period")
        
        # Calculate summary
        total_pnl = data[-1].cumulative_pnl if data else 0
        total_pnl_percent = data[-1].pnl_percent if data else 0
        
        # Find best and worst days
        sorted_by_pnl = sorted(data, key=lambda x: x.pnl, reverse=True)
        best_day = sorted_by_pnl[0] if sorted_by_pnl else None
        worst_day = sorted_by_pnl[-1] if sorted_by_pnl else None
        
        # Calculate volatility (simplified)
        pnl_values = [d.pnl for d in data]
        mean_pnl = sum(pnl_values) / len(pnl_values) if pnl_values else 0
        variance = sum((x - mean_pnl) ** 2 for x in pnl_values) / len(pnl_values) if pnl_values else 0
        volatility = (variance ** 0.5) / 10000  # Convert to 万
        
        # Mock benchmark comparison
        benchmark_comparison = {
            "portfolio_return": total_pnl_percent,
            "benchmark_return": total_pnl_percent * 0.85,  # Slightly underperform
            "alpha": total_pnl_percent * 0.15,
            "beta": 0.95
        }
        
        summary = PnLSummary(
            total_pnl=total_pnl,
            total_pnl_formatted=f"+{total_pnl/10000:.1f}万" if total_pnl >= 0 else f"{total_pnl/10000:.1f}万",
            total_pnl_percent=total_pnl_percent,
            best_day=best_day,
            worst_day=worst_day,
            volatility=volatility,
            sharpe_ratio=1.2 if total_pnl > 0 else -0.5
        )
        
        result = PnLHistoryResponse(
            period=period.value,
            start_date=data[0].date if data else date.today().isoformat(),
            end_date=data[-1].date if data else date.today().isoformat(),
            data=data,
            summary=summary,
            benchmark_comparison=benchmark_comparison
        )
        
        return ApiResponse(code=200, message="success", data=result)
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to get PnL history: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/pnl-history/daily")
async def get_daily_pnl(
    start_date: Optional[str] = Query(None, description="Start date (YYYY-MM-DD)"),
    end_date: Optional[str] = Query(None, description="End date (YYYY-MM-DD)")
):
    """Get daily PnL for custom date range."""
    try:
        start = datetime.strptime(start_date, "%Y-%m-%d").date() if start_date else date.today() - timedelta(days=30)
        end = datetime.strptime(end_date, "%Y-%m-%d").date() if end_date else date.today()
        
        days = (end - start).days
        data = generate_pnl_history("custom", days)
        
        # Filter to date range
        data = [d for d in data if start.isoformat() <= d.date <= end.isoformat()]
        
        return ApiResponse(code=200, message="success", data=data)
        
    except Exception as e:
        logger.error(f"Failed to get daily PnL: {e}")
        raise HTTPException(status_code=500, detail=str(e))


# ============================================================================
# API Endpoints - Issue #209
# ============================================================================

@router.get("/sector-allocation", response_model=ApiResponse[SectorAllocationResponse])
async def get_sector_allocation(
    user_id: Optional[int] = Query(None, description="User ID (for admin/multi-user)"),
    min_percent: float = Query(0.5, ge=0, le=100, description="Minimum percentage to include (others grouped)")
):
    """Get portfolio sector allocation distribution.
    
    Issue #209: Returns sector allocation for portfolio diversification analysis.
    Aggregates holdings by industry/sector with percentage breakdown.
    
    Args:
        user_id: Optional user ID for multi-user systems
        min_percent: Minimum percentage for individual sector display
        
    Returns:
        SectorAllocationResponse with sector breakdown and diversification metrics.
    """
    try:
        # Mock portfolio holdings by sector
        mock_sectors = [
            {"name": "食品饮料", "value": 850000, "stocks": [{"symbol": "600519", "name": "贵州茅台", "value": 500000}, {"symbol": "000858", "name": "五粮液", "value": 350000}]},
            {"name": "医药生物", "value": 620000, "stocks": [{"symbol": "600276", "name": "恒瑞医药", "value": 320000}, {"symbol": "300760", "name": "迈瑞医疗", "value": 300000}]},
            {"name": "电子", "value": 480000, "stocks": [{"symbol": "002594", "name": "比亚迪", "value": 280000}, {"symbol": "300750", "name": "宁德时代", "value": 200000}]},
            {"name": "银行", "value": 320000, "stocks": [{"symbol": "600036", "name": "招商银行", "value": 200000}, {"symbol": "601398", "name": "工商银行", "value": 120000}]},
            {"name": "电力设备", "value": 210000, "stocks": [{"symbol": "601012", "name": "隆基绿能", "value": 150000}, {"symbol": "300274", "name": "阳光电源", "value": 60000}]},
            {"name": "汽车", "value": 180000, "stocks": [{"symbol": "000625", "name": "长安汽车", "value": 100000}, {"symbol": "601633", "name": "长城汽车", "value": 80000}]},
            {"name": "计算机", "value": 120000, "stocks": [{"symbol": "000938", "name": "紫光股份", "value": 70000}, {"symbol": "600570", "name": "恒生电子", "value": 50000}]},
            {"name": "家用电器", "value": 85000, "stocks": [{"symbol": "000333", "name": "美的集团", "value": 85000}]},
            {"name": "化工", "value": 65000, "stocks": [{"symbol": "600309", "name": "万华化学", "value": 65000}]},
            {"name": "有色金属", "value": 45000, "stocks": [{"symbol": "601899", "name": "紫金矿业", "value": 45000}]},
        ]
        
        # Calculate total
        total_value = sum(s["value"] for s in mock_sectors)
        cash_value = 500000  # 50万现金
        total_with_cash = total_value + cash_value
        
        # Build sector allocations
        sectors = []
        others_value = 0
        others_stocks = []
        
        for sector_data in mock_sectors:
            percent = (sector_data["value"] / total_with_cash) * 100
            
            if percent >= min_percent:
                sectors.append(SectorAllocationItem(
                    name=sector_data["name"],
                    code=sector_data["name"][:2].upper(),
                    value=sector_data["value"],
                    value_formatted=f"{sector_data['value']/10000:.1f}万",
                    percent=round(percent, 1),
                    color=IndustryMapping.get_color(sector_data["name"]),
                    stock_count=len(sector_data["stocks"]),
                    stocks=sector_data["stocks"]
                ))
            else:
                others_value += sector_data["value"]
                others_stocks.extend(sector_data["stocks"])
        
        # Add "Others" category if needed
        if others_value > 0:
            others_percent = (others_value / total_with_cash) * 100
            sectors.append(SectorAllocationItem(
                name="其他",
                code="OTHERS",
                value=others_value,
                value_formatted=f"{others_value/10000:.1f}万",
                percent=round(others_percent, 1),
                color=IndustryMapping.DEFAULT_COLOR,
                stock_count=len(others_stocks),
                stocks=others_stocks[:3]  # Show top 3
            ))
        
        # Add cash position
        cash_percent = (cash_value / total_with_cash) * 100
        sectors.append(SectorAllocationItem(
            name="现金",
            code="CASH",
            value=cash_value,
            value_formatted=f"{cash_value/10000:.1f}万",
            percent=round(cash_percent, 1),
            color="#64748B",
            stock_count=0,
            stocks=[]
        ))
        
        # Sort by value descending
        sectors.sort(key=lambda x: x.value, reverse=True)
        
        # Calculate diversification metrics
        # Herfindahl-Hirschman Index (HHI) based on sector concentration
        hhi = sum((s.percent / 100) ** 2 for s in sectors)
        diversification_score = (1 - hhi) * 100  # 0-100 score
        
        # Determine concentration risk
        max_sector_percent = max(s.percent for s in sectors)
        if max_sector_percent > 40:
            risk_level = "high"
        elif max_sector_percent > 25:
            risk_level = "medium"
        else:
            risk_level = "low"
        
        result = SectorAllocationResponse(
            total_value=total_with_cash,
            total_value_formatted=f"{total_with_cash/10000:.1f}万",
            cash_value=cash_value,
            cash_percent=round(cash_percent, 1),
            sectors=sectors,
            diversification_score=round(diversification_score, 1),
            top_heavy_risk=risk_level
        )
        
        return ApiResponse(code=200, message="success", data=result)
        
    except Exception as e:
        logger.error(f"Failed to get sector allocation: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/sector-allocation/trend")
async def get_sector_allocation_trend(
    months: int = Query(6, ge=1, le=24, description="Number of months to analyze")
):
    """Get sector allocation trend over time."""
    try:
        trends = []
        today = date.today()
        
        for i in range(months):
            month_date = today - timedelta(days=i*30)
            trends.append({
                "month": month_date.strftime("%Y-%m"),
                "top_sector": "食品饮料" if i % 3 == 0 else "医药生物" if i % 3 == 1 else "电子",
                "top_sector_percent": 25 + (i % 5),
            })
        
        return ApiResponse(code=200, message="success", data=trends)
        
    except Exception as e:
        logger.error(f"Failed to get sector trend: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/holdings")
async def get_portfolio_holdings(
    sort_by: str = Query("value", description="Sort by: value, pnl, weight"),
    limit: int = Query(50, ge=1, le=100, description="Number of holdings to return")
):
    """Get detailed portfolio holdings."""
    try:
        # Mock holdings data
        holdings = [
            {"symbol": "600519", "name": "贵州茅台", "sector": "食品饮料", "shares": 200, "cost": 150000, "value": 356000, "pnl": 206000, "pnl_percent": 137.3, "weight": 12.5},
            {"symbol": "000858", "name": "五粮液", "sector": "食品饮料", "shares": 1500, "cost": 280000, "value": 390000, "pnl": 110000, "pnl_percent": 39.3, "weight": 13.7},
            {"symbol": "300750", "name": "宁德时代", "sector": "电力设备", "shares": 800, "cost": 160000, "value": 156800, "pnl": -3200, "pnl_percent": -2.0, "weight": 5.5},
            {"symbol": "600036", "name": "招商银行", "sector": "银行", "shares": 5000, "cost": 180000, "value": 200000, "pnl": 20000, "pnl_percent": 11.1, "weight": 7.0},
            {"symbol": "002594", "name": "比亚迪", "sector": "汽车", "shares": 600, "cost": 156000, "value": 168000, "pnl": 12000, "pnl_percent": 7.7, "weight": 5.9},
        ]
        
        # Sort
        if sort_by == "pnl":
            holdings.sort(key=lambda x: x["pnl"], reverse=True)
        elif sort_by == "weight":
            holdings.sort(key=lambda x: x["weight"], reverse=True)
        
        return ApiResponse(code=200, message="success", data=holdings[:limit])
        
    except Exception as e:
        logger.error(f"Failed to get holdings: {e}")
        raise HTTPException(status_code=500, detail=str(e))
